pkg/bytecode/symbol.go

```diff
@@ -27,20 +41,61 @@ func NewSymbolTable() *SymbolTable {
 	}
 }
 
+// Push nests a new symbol table under s. The index for the nested table starts
+// at the index of the outer symbol table if the outer symbol table is not the
+// global symbol table. Otherwise the index starts at zero.
+func (s *SymbolTable) Push() *SymbolTable {
+	index := s.index
+	// If our outer symbol table is the global symbol table, we do
+	// not count its index as globals are stored in a separate space.
+	if s.outer == nil {
+		index = 0
+	}
+	return &SymbolTable{
+		store: make(map[string]Symbol),
+		outer: s,
+		index: index,
+	}
+}
+
+// Pop returns the outer symbol table of s and updates the outer symbol table's
+// nested max index. This accumulates the maximum index of all nested scopes
+// into the parent scope so we know how much storage is needed for all the
+// variables of a scope and its children.
+func (s *SymbolTable) Pop() *SymbolTable {
+	if s.outer == nil {
+		return s
+	}
+	s.outer.nestedMaxIndex = max(s.outer.nestedMaxIndex, s.nestedMaxIndex+s.index)
+	return s.outer
+}
+
 // Define adds a symbol definition to the table or returns an
 // already defined symbol with the same name.
 func (s *SymbolTable) Define(name string) Symbol {
 	if existing, found := s.store[name]; found {
 		return existing
 	}
-	symbol := Symbol{Name: name, Index: len(s.store), Scope: GlobalScope}
+	symbol := Symbol{Name: name, Index: s.index}
+	s.index++
+	if s.outer == nil {
+		symbol.Scope = GlobalScope
+	} else {
+		symbol.Scope = LocalScope
```

Comment:
So just to be clear: local means any block scope or function level scoped variable, correct?

Reply:
Yes - anything that is not at the top-level of the program.